modules/packages/Futures.chpl

```diff
@@ -0,0 +1,240 @@
+
+   writeln(F.get() == (3 * ((2 * X) + 7))); // prints "true"
+
+.. code-block:: chapel
+
+   use Futures;
+
+   config const X = 1234;
+
+   var F = async(lambda(x:int) { return x:string; }, X)
+     .andThen(lambda(x:string) { return(x.length); });
+
+   writeln(F.get()); // prints "4"
+
+ */
+
+module Futures {
+
+  use Reflection;
+  use RefCount;
+
+  pragma "no doc"
+  class FutureClass: RefCountBase {
+
+    type retType;
+
+    var value: single retType;
+
+    proc FutureClass(type retType) {
+    }
+
+  } // class FutureClass
+
+  /*
+    A container that can store the result of an asynchronous operation,
+    which can be retrieved when the result is ready.
+
+    A future is not created directly. Instead, one is created by calling the
+    :proc:`async()` function or the :proc:`Future.andThen()` method on
+    an already-existing future.
+   */
+  record Future {
+
+    /*
+      The return type of the future.
+     */
+    type retType;
+
+    pragma "no doc"
+    var classRef: FutureClass(retType);
+
+    pragma "no doc"
+    proc Future(type retType) {
+      acquire(new FutureClass(retType));
+    }
+
+    pragma "no doc"
+    proc ~Future() {
+      release();
+    }
+
+    /*
+      Get the result of a future, blocking until it is available.
+     */
+    proc get(): retType {
+      return classRef.value.readFF();
+    }
+
+    /*
+      Test whether the result of the future is available.
+     */
+    proc isReady(): bool {
+      return classRef.value.isFull;
+    }
+
+    /*
+      Asynchronously execute a function as a continuation of the future.
+
+      The function argument `taskFn` must take a single argument of type
+      `retType` (i.e., the return type of the parent future) and will be
+      executed when the parent future's value is available.
+
+      :arg taskFn: The function to invoke as a continuation.
+      :returns: A future of the return type of `taskFn`
+     */
+    proc andThen(taskFn) {
+      /*
+      if !canResolveMethod(taskFn, "this", retType) then
+        compilerError("Mismatch in async task function and arguments");
+      */
+      var f: Future(taskFn.retType);
+      begin f.classRef.value.writeEF(taskFn(this.get()));
+      return f;
+    }
+
+    pragma "no doc"
+    proc acquire(newRef: FutureClass) {
+      classRef = newRef;
+      classRef.incRefCount();
+    }
+
+    pragma "no doc"
+    proc acquire() {
+      classRef.incRefCount();
+    }
+
+    pragma "no doc"
+    proc release() {
+      var rc = classRef.decRefCount();
+      if rc == 1 {
+        delete classRef;
+        classRef = nil;
+      }
+    }
+
+  } // record Future
+
+  pragma "no doc"
+  pragma "init copy fn"
+  proc chpl__initCopy(x: Future) {
+    x.acquire();
+    return x;
+  }
+
+  pragma "no doc"
+  pragma "auto copy fn"
+  proc chpl__autoCopy(x: Future) {
+    x.acquire();
+    return x;
+  }
+
+  pragma "no doc"
+  proc =(ref lhs: Future, rhs: Future) {
+  if lhs.classRef != nil then
```

Comment:
> The normal pattern for reference counting...

Good point about the release-acquire order here. (If it's a bug here, which I agree it is, then it's also a bug in the ZeroMQ module.)

> ...you'd have to write something like...

I have intentionally written tests like that. I think of the declaration-without-initialization pattern as something like an "incomplete future" (I need a better term here). In your example, calling `future.get()` (before the `async()` call) would block forever.

I can add some runtime logic that notes whether a future is incomplete the sense here and calls `halt()` if `get()` is called on an incomplete future. (I'd update this to use Chapel's future error throwing/handling when that's available.) I think the C++ analog to "incomplete future" as I'm using it here is whether the future refers to any shared state, which can be checked via [`std::future::valid()`](http://en.cppreference.com/w/cpp/thread/future/valid).

> IIRC, the Rice group thought that you shouldn't be able to assign futures. [...]

One thing 2f1b857 (vs 1fa8ffe) was meant to change is to explicitly allow assignment of futures, particularly for the declaration-without-initialization pattern. Otherwise, one cannot create an array of (incomplete) futures and create acyclic dependencies (e.g., for a 2D wavefront propagation). If I add runtime checks for `get()` on a incomplete futures, I could add similar logic for assignment. I think this also relates to whether the future is a record-wrapped, reference-counted class or just a plain record.